### PR TITLE
feat: Add `BuildArgs` to `build_program`

### PR DESCRIPTION
**Pull request by ratankaliani**
In certain cases, we want to conditionally build the binary with some custom flags (ex. Sovereign). Move the logic for building a program into `sp1-build`, and call this logic from both `sp1-helper` and `sp1-cli`.

Add the following feature flags to `BuildArgs`
- `binary`: Specific binary to target in the program directory.
- `elf-name`: Name of the resulting ELF.
- `features`: Features to compile the program with.
- `output_directory`: Directory to output the built ELF to.

Additonally, adds support for `SP1_SKIP_PROGRAM_BUILD`, which is useful in CI for skipping building the program so that you don't need the Succinct Rust toolchain installed.

### CODE REVIEW

**Review by ctian1 — APPROVED** [review]

File: build/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+    // Add the `cargo prove build` arguments to the child process command, while ignoring
+    // `--docker`, as we already know it's unused in this conditional.
+    add_cargo_prove_build_args(&mut cargo_args, args.clone(), true);
+
+    // Ensure the Cargo.lock doesn't update.
+    cargo_args.push("--locked".to_string());
+
+    command
+        .current_dir(program_dir.clone())
+        .env("RUSTUP_TOOLCHAIN", "succinct")
```
Comment (ratankaliani) [inline build/src/lib.rs:145]: I'm confused why this isn't activating...

Comment (ratankaliani) [reply build/src/lib.rs:145]: I needed to strip the `RUSTC` environment variable in the command, as it was inheriting the `RUSTC` from the parent build process.

File: build/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        build_args.push(args.binary.clone());
+    }
+
+    if !args.features.is_empty() {
+        build_args.push("--features".to_string());
+        build_args.push(args.features.join(","));
+    }
+
+    // Ensure the Cargo.lock doesn't update.
+    build_args.push("--locked".to_string());
```
[the remark this answers is not in the document]
Comment (ratankaliani) [reply build/src/lib.rs]: Added `locked` flag, by default this flag is off now.

File: build/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    pub features: Vec<String>,
+    #[clap(long, action, help = "Ignore Rust version check.")]
+    pub ignore_rust_version: bool,
+    #[clap(
+        long,
+        action,
+        help = "If building a binary, specify the name.",
+        default_value = ""
+    )]
+    pub binary: String,
```
[the remark this answers is not in the document]
Comment (ratankaliani) [reply build/src/lib.rs:42]: Supported with an alias now.

File: build/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    pub tag: String,
+    #[clap(long, action, value_delimiter = ',', help = "Build with features.")]
+    pub features: Vec<String>,
+    #[clap(long, action, help = "Ignore Rust version check.")]
+    pub ignore_rust_version: bool,
+    #[clap(
+        long,
+        action,
+        help = "If building a binary, specify the name.",
+        default_value = ""
```
[the remark this answers is not in the document]
Comment (ratankaliani) [reply build/src/lib.rs:40]: `binary` here refers to the Rust binary that should be run, not the toolchain type.

File: helper/src/lib.rs [hunk 5]
```diff
@@ -1,56 +1,31 @@
...
-    let mut metadata_cmd = cargo_metadata::MetadataCommand::new();
-    let metadata = metadata_cmd.manifest_path(metadata_file).exec().unwrap();
-    let root_package = metadata.root_package();
-    let root_package_name = root_package
-        .as_ref()
-        .map(|p| p.name.as_str())
-        .unwrap_or("Program");
+/// Re-run the cargo command if the Cargo.toml or Cargo.lock file changes.
+fn cargo_rerun_if_changed(metadata: &Metadata) {
+    // Re-run the build script if Cargo.lock changes or any dependency changes.
```
[the remark this answers is not in the document]
Comment (ratankaliani) [reply helper/src/lib.rs]: Sounds good, added. Here, I ensure that bin and src exist so that metadata caching works correctly.

**Pull request merged**